modules/auxiliary/gather/enum_dns.rb

```diff
@@ -1,483 +1,420 @@
+      '_test._tcp.',  '_sips._tcp.', '_sip._udp.', '_sip._tcp.',
+      '_aix._tcp.',  '_aix._tcp.', '_finger._tcp.', '_ftp._tcp.',
+      '_http._tcp.', '_nntp._tcp.', '_telnet._tcp.', '_whois._tcp.',
+      '_h323cs._tcp.',  '_h323cs._udp.', '_h323be._tcp.', '_h323be._udp.',
+      '_h323ls._tcp.', '_h323ls._udp.',  '_sipinternal._tcp.',
+      '_sipinternaltls._tcp.', '_sip._tls.', '_sipfederationtls._tcp.',
+      '_jabber._tcp.', '_xmpp-server._tcp.', '_xmpp-client._tcp.',
+      '_imap._tcp.', '_certificates._tcp.',  '_crls._tcp.', '_pgpkeys._tcp.',
+      '_pgprevokations._tcp.', '_cmp._tcp.', '_svcp._tcp.', '_crl._tcp.',
+      '_ocsp._tcp.', '_PKIXREP._tcp.', '_smtp._tcp.', '_hkp._tcp.',
+      '_hkps._tcp.', '_jabber._udp.',  '_xmpp-server._udp.',
+      '_xmpp-client._udp.', '_jabber-client._tcp.', '_jabber-client._udp.']
 
-    if(datastore['ENUM_BRT'])
-      if wldcrd and datastore['STOP_WLDCRD']
-        print_error("Wildcard record found!")
-      else
-        dnsbrute(datastore['DOMAIN'],datastore['WORDLIST'],datastore['NS'])
+    records = []
+    srvs.each do |srv|
+      vprint_status("#{srv}#{domain}")
+      resp = dns_query("#{srv}#{domain}", Net::DNS::SRV)
+      next if resp.blank? || resp.answer.blank?
+      resp.answer.each do |r|
+        next if r.type == Net::DNS::RR::CNAME
+        report_note(
+          host: domain,
+          proto: 'udp',
+          sname: r.host,
+          port: '53',
+          type: 'ENUM_SRV',
+          data: "#{r.priority}")
+        vprint_good("Host: #{r.host} Port: #{r.port} Priority: #{r.priority}")
       end
     end
+    return if records.none?
+    path = store_loot(
+      'ENUM_SRV',
+      'text/plain',
+      domain,
+      records,
+      'ENUM_SRV')
+    print_good('Saved file to: ' + path)
+    records
+  end
 
-    if(datastore['ENUM_IP6'])
-      if wldcrd and datastore['STOP_WLDCRD']
-        print_status("Wildcard Record Found!")
-      else
-        bruteipv6(datastore['DOMAIN'],datastore['WORDLIST'],datastore['NS'])
-      end
-    end
+  def axfr(domain)
+    nameservers = get_ns(domain)
+    return if nameservers.blank?
 
-    if(datastore['ENUM_AXFR'])
-      axfr(datastore['DOMAIN'],datastore['NS'])
-    end
+    records = []
+    nameservers.each do |nameserver|
+      dns = Net::DNS::Resolver.new
+      dns.use_tcp = datastore['TCP_DNS']
+      dns.udp_timeout = datastore['TIMEOUT']
+      dns.retry_number = datastore['RETRY']
+      dns.retry_interval = datastore['RETRY_INTERVAL']
 
-    if(datastore['ENUM_SRV'])
-      srvqry(datastore['DOMAIN'],datastore['NS'])
-    end
+      next if nameserver.blank?
+      ns = get_a(nameserver)
+      next if ns.blank?
 
-    if(datastore['ENUM_RVL'] and datastore['IPRANGE'] and not datastore['IPRANGE'].empty?)
-      reverselkp(datastore['IPRANGE'],datastore['NS'])
+      ns.each do |r|
+        begin
+          dns.nameservers -= dns.nameservers
+          dns.nameservers = "#{r}"
```

Comment:
```
msf auxiliary(enum_dns) > show options

Module options (auxiliary/gather/enum_dns):

   Name         Current Setting                                                             Required  Description
   ----         ---------------                                                             --------  -----------
   DOMAIN       metasploit.com                                                              yes       The target domain
   ENUM_AXFR    true                                                                        yes       Initiate a zone transfer against each NS record
   ENUM_BRT     false                                                                       yes       Brute force subdomains and hostnames via the supplied wordlist
   ENUM_SRV     false                                                                       yes       Enumerate the most common SRV records
   ENUM_STD     true                                                                        yes       Enumerate standard record types (A,MX,NS,TXT and SOA)
   ENUM_TLD     false                                                                       yes       Perform a TLD expansion by replacing the TLD with the IANA TLD list
   NS           8.8.8.8                                                                     no        Nameserver for query
   STOP_WLDCRD  false                                                                       yes       Stops bruteforce enumeration if wildcard resolution is detected
   THREADS      1                                                                           no        Threads for ENUM_BRT
   WORDLIST     /Users/Open-Security/Code/metasploit-framework/data/wordlists/namelist.txt  no        Wordlist of subdomains
```

1. Set **```NS```** to 8.8.8.8.
2. ```dns.nameservers -= dns.nameservers``` is necessary for users. Please compare demo 1 and demo 2. If we remove the line, dns.nameservers will  be not my choice.


**Demo 1** - remove the line

When we try to zone transfer target, ```dns.nameservers``` is filled by previous dns server, but not nameservers from metasploit.com. 
```
From: /Users/Open-Security/Code/metasploit-framework/modules/auxiliary/gather/enum_dns.rb @ line 396 Msf::Modules::Mod617578696c696172792f6761746865722f656e756d5f646e73::Metasploit3#axfr:

    391: 
    392:       ns.each do |r|
    393:         begin
    394:           # dns.nameservers -= dns.nameservers
    395:           dns.nameservers = "#{r}"
 => 396:           require 'pry'
    397:           binding.pry
    398:           zone = dns.axfr(domain)
    399:         rescue ::Rex::ConnectionError

[1] pry(#<Msf::Modules::Mod617578696c696172792f6761746865722f656e756d5f646e73::Metasploit3>)> dns.nameservers
=> ["223.5.5.5"]
[2] pry(#<Msf::Modules::Mod617578696c696172792f6761746865722f656e756d5f646e73::Metasploit3>)> r
=> "205.251.197.161"
```

**Demo 2**

```
From: /Users/Open-Security/Code/metasploit-framework/modules/auxiliary/gather/enum_dns.rb @ line 397 Msf::Modules::Mod617578696c696172792f6761746865722f656e756d5f646e73::Metasploit3#axfr:

    392:       ns.each do |r|
    393:         begin
    394:           dns.nameservers -= dns.nameservers
    395:           dns.nameservers = "#{r}"
    396:           require 'pry'
 => 397:           binding.pry
    398:           zone = dns.axfr(domain)
    399:         rescue ::Rex::ConnectionError


[1] pry(#<Msf::Modules::Mod617578696c696172792f6761746865722f656e756d5f646e73::Metasploit3>)> dns.nameservers
=> ["205.251.197.161"]
```